common/src/main/java/org/keycloak/common/crypto/CryptoIntegration.java

```diff
@@ -0,0 +1,52 @@
+package org.keycloak.common.crypto;
+
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.BouncyIntegration;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class CryptoIntegration {
+
+    protected static final Logger logger = Logger.getLogger(CryptoIntegration.class);
+
+    private static volatile AtomicReference<CryptoProvider> securityProvider = new AtomicReference<>();
+
+    public static void init(ClassLoader classLoader) {
+        securityProvider.set(detectProvider(classLoader));
+        logger.debugv("BouncyCastle provider: {0}", BouncyIntegration.PROVIDER);
+    }
+
+    public static CryptoProvider getProvider() {
+        CryptoProvider cryptoProvider = securityProvider.get();
+        if (cryptoProvider == null) {
+            securityProvider.compareAndSet(null, detectProvider(CryptoIntegration.class.getClassLoader()));
+            cryptoProvider = securityProvider.get();
+        }
+        return cryptoProvider;
```

Comment:
The simple static init does not work due the fact that classloader needs to be passed. Note that classloader of `CryptoIntegration` itself is `keycloak-common` module, which does not have access to crypto modules. In the Wildfly, there is classloader of `keycloak-services` passed here from `KeycloakApplication` during initialization (or classloader of `keycloak-adapter-core` is passed in case of adapters).

I can possibly enforce `init` called before any call to `getProvider` . The only reason why it was not yet done are the unit tests. So in that case the getProvider method would look like this ( I would prefer this instead of just `securityProvider.get()` just to avoid NullPointerException, but rather `IllegalStateException` with some cleaner description of what is going on):
```
    public static CryptoProvider getProvider() {
        CryptoProvider cryptoProvider = securityProvider.get();
        if (cryptoProvider == null) {
            throw new IllegalStateException("Illegal state. Please init first before obtaining provider");
        }
        return cryptoProvider;
    }
```

WDYT?